### ISSUE TITLE
Lint documentation examples

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#2271]
Split from [#2257 (comment)](https://github.com/eslint/eslint/pull/2257#discussion_r28074602).

With file loaders, we can now lint the example code in documentation files. At the very least, this should catch syntax errors. I think it would be cool if we used configuration comments and actually applied the rule to the valid examples. (Not sure how we would only check the invalid ones for syntax; that'll take some thought.)

**Comment by IanVS** [comment]
I'll take a crack at this.

**Comment by IanVS** [comment]
Here are a list of the current rules and status:

* [X] accessor-pairs
* [X] array-bracket-spacing
* [X] arrow-parens
* [X] arrow-spacing
* [X] block-scoped-var
* [ ] block-spacing
* [X] brace-style
* [X] callback-return
* [X] camelcase
* [X] comma-dangle
* [X] comma-spacing
* [X] comma-style
* [X] complexity
* [X] computed-property-spacing
* [X] consistent-return
* [X] consistent-this
* [X] constructor-super
* [X] curly
* [X] default-case
* [X] dot-location
* [X] dot-notation
* [X] eol-last
* [X] eqeqeq
* [X] func-names
* [X] func-style
* [X] generator-star-spacing
* [X] guard-for-in
* [X] handle-callback-err
* [X] id-length
* [X] id-match
* [X] indent
* [X] init-declarations
* [X] key-spacing
* [X] linebreak-style
* [X] lines-around-comment
* [X] max-depth
* [X] max-len
* [X] max-nested-callbacks
* [X] max-params
* [X] max-statements
* [X] new-cap
* [X] newline-after-var
* [X] new-parens
* [X] no-alert
* [X] no-array-constructor
* [X] no-bitwise
* [X] no-caller
* [X] no-catch-shadow
* [X] no-class-assign
* [X] no-cond-assign
* [X] no-console
* [X] no-constant-condition
* [X] no-const-assign
* [X] no-continue
* [X] no-control-regex
* [ ] no-debugger
* [X] no-delete-var
* [X] no-div-regex
* [X] no-dupe-args
* [X] no-dupe-class-members
* [X] no-dupe-keys
* [X] no-duplicate-case
* [X] no-else-return
* [X] no-empty
* [X] no-empty-character-class
* [X] no-empty-label
* [X] no-eq-null
* [X] no-eval
* [X] no-ex-assign
* [X] no-extend-native
* [X] no-extra-bind
* [X] no-extra-boolean-cast
* [X] no-extra-parens
* [X] no-extra-semi
* [ ] no-fallthrough
* [ ] no-floating-decimal
* [ ] no-func-assign
* [ ] no-implicit-coercion
* [ ] no-implied-eval
* [ ] no-inline-comments
* [ ] no-inner-declarations
* [ ] no-invalid-regexp
* [ ] no-invalid-this
* [ ] no-irregular-whitespace
* [ ] no-iterator
* [ ] no-labels
* [ ] no-label-var
* [ ] no-lone-blocks
* [ ] no-lonely-if
* [ ] no-loop-func
* [ ] no-mixed-requires
* [ ] no-mixed-spaces-and-tabs
* [ ] no-multiple-empty-lines
* [ ] no-multi-spaces
* [ ] no-multi-str
* [ ] no-native-reassign
* [ ] no-negated-in-lhs
* [ ] no-nested-ternary
* [ ] no-new
* [ ] no-new-func
* [ ] no-new-object
* [ ] no-new-require
* [ ] no-new-wrappers
* [ ] no-obj-calls
* [ ] no-octal
* [ ] no-octal-escape
* [ ] no-param-reassign
* [ ] no-path-concat
* [ ] no-plusplus
* [ ] no-process-env
* [ ] no-process-exit
* [ ] no-proto
* [ ] no-redeclare
* [ ] no-regex-spaces
* [ ] no-restricted-modules
* [ ] no-return-assign
* [ ] no-script-url
* [ ] no-self-compare
* [ ] no-sequences
* [ ] no-shadow
* [ ] no-shadow-restricted-names
* [ ] no-spaced-func
* [ ] no-sparse-arrays
* [ ] no-sync
* [ ] no-ternary
* [ ] no-this-before-super
* [ ] no-throw-literal
* [ ] no-trailing-spaces
* [ ] no-undef
* [ ] no-undefined
* [ ] no-undef-init
* [ ] no-underscore-dangle
* [ ] no-unexpected-multiline
* [ ] no-unneeded-ternary
* [ ] no-unreachable
* [ ] no-unused-expressions
* [ ] no-unused-vars
* [ ] no-use-before-define
* [ ] no-useless-call
* [ ] no-var
* [ ] no-void
* [ ] no-warning-comments
* [ ] no-with
* [ ] object-curly-spacing
* [ ] object-shorthand
* [ ] one-var
* [ ] operator-assignment
* [ ] operator-linebreak
* [ ] padded-blocks
* [ ] prefer-arrow-callback
* [ ] prefer-const
* [ ] prefer-reflect
* [ ] prefer-spread
* [ ] prefer-template
* [ ] quote-props
* [ ] quotes
* [ ] radix
* [ ] require-yield
* [ ] semi
* [ ] semi-spacing
* [ ] sort-vars
* [ ] space-after-keywords
* [ ] space-before-blocks
* [ ] space-before-function-paren
* [ ] spaced-comment
* [ ] space-infix-ops
* [ ] space-in-parens
* [ ] space-return-throw-case
* [ ] space-unary-ops
* [ ] strict
* [ ] use-isnan
* [ ] valid-jsdoc
* [ ] valid-typeof
* [ ] vars-on-top
* [ ] wrap-iife
* [ ] wrap-regex
* [ ] yoda

**Comment by nzakas** [comment]
Wowza.

**Comment by IanVS** [comment]
:dizzy: :zzz:

**Comment by IanVS** [comment]
Second PR in progress: https://github.com/eslint/eslint/pull/3574

**Comment by IanVS** [comment]
The way the markdown linter plugin currently works, all expected errors must be provided on the line for which they are reported.  This causes some readability problems when there are multiple errors for the same line.  After https://github.com/eslint/eslint-plugin-markdown/issues/9 is solved, it will be possible to shorten this to simply the number of expected errors.

Tracking the list of rules with more than one error on a line so they can be cleaned up later:
* object-curly-spacing

**Comment by IanVS** [comment]
I've finally hit a rule for which I cannot provide expected errors the way I've been doing.  That is `valid-jsdoc`, because as soon as I add an `/*error ...*/` comment on the line with the error, the comment is no longer valid jsdoc and stops being considered by the rule.  For now, I'll set the rule to warning, but will likely need a better long-term solution.

**Comment by IanVS** [comment]
`no-warning-comments` is waiting on https://github.com/eslint/eslint/issues/3619.

**Comment by nzakas** [comment]
How about using line comments instead of block comments?

**Comment by IanVS** [comment]
> How about using line comments instead of block comments?

That works as long as there's only one error.  Or once https://github.com/eslint/eslint-plugin-markdown/issues/9 is implemented, this could be just the ticket.

**Comment by IanVS** [comment]
(broken out from main tracking comment above)

Some rule examples require ecmaFeatures: { modules: true } and others require modules: false. Until a better solution can be found (like functionality for modules similar to https://github.com/eslint/eslint/pull/3314), I'm removing the info string from the rules which require modules to be turned off. These are:

* no-delete-var
* no-dupe-args
* no-lone-blocks (a solution to #2134 would also work here)
* no-octal-escape
* no-octal
* no-restricted-syntax
* no-sequences
* no-shadow-restricted-names
* no-with
* prefer-reflect
* global

**Comment by IanVS** [comment]
These are the rules mentioned above which have issues:

* [ ] indent (DavidAnson/markdownlint#5)
* [ ] no-trailing-spaces (DavidAnson/markdownlint#5)
* [ ] no-void (purposeful syntax error, no way to mark as expected)
* [ ] no-warning-comments (#3619 - fixed)
* [ ] valid-jsdoc (eslint/eslint-plugin-markdown#9 - fixed)
* [ ] no-delete-var (turn off modules/strict mode)
* [ ] no-dupe-args (turn off modules/strict mode)
* [ ] no-lone-blocks (turn off modules/strict mode)
* [ ] no-octal-escape (turn off modules/strict mode)
* [ ] no-octal (turn off modules/strict mode)
* [ ] no-restricted-syntax (turn off modules/strict mode)
* [ ] no-sequences (turn off modules/strict mode)
* [ ] no-shadow-restricted-names (turn off modules/strict mode)
* [ ] no-with (turn off modules/strict mode)
* [ ] prefer-reflect (turn off modules/strict mode)
* [ ] global (turn off modules/strict mode)

**Comment by gyandeeps** [comment]
@IanVS Can we close this based on https://github.com/eslint/eslint/issues/4104?

**Comment by IanVS** [comment]
It would still be nice to find a way to lint the rule examples.   Some discussion has been happening in https://github.com/eslint/eslint-plugin-markdown/issues/46 as to how we might accomplish this.

**Comment by ilyavolodin** [comment]
I have what might be somewhat of a radical idea, but what if we stop adding examples to documentation and instead, generate them from test cases for rules? We can provide placeholders for where they are supposed to go into. This way we know that example are always going to be valid.

**Comment by platinumazure** [comment]
@ilyavolodin Awesome idea, :+1: from me, but sometimes there are a lot of test cases. How do you propose avoiding information overload?

**Comment by btmills** [comment]
There's probably a way to cross-reference specific tests...

```md
The following patterns are not considered problems:

[](#eslint-test some-valid-example)
[](#eslint-test this-one-too)

The following patterns are considered problems:

[](#eslint-test first-bad-example)
[](#eslint-test another-bad-example)
```

```js
var testCases = module.exports = {
    valid: {
        "anonymousTest();",
        { code: "namedTest();", name: "some-valid-example" }
    },
    invalid: {
        { /* ... */ name: "another-bad-example" }
    }
};

ruleTester.run("rule-name", rule, testCases);
```

It would require refactoring the tests to export the test cases, but I suppose there could be a step during site generation that could extract those.

**Comment by nzakas** [comment]
@btmills @IanVS any updates on this? Still planning to move forward?

**Comment by IanVS** [comment]
I have no immediate plans to pursue this myself, no.

**Comment by btmills** [comment]
There's an open issue (https://github.com/eslint/eslint-plugin-markdown/issues/46) discussing how to support expected errors in linter examples. How about closing this, tracking possible ideas over there, and if we can figure that out, we can open a new issue here?

**Comment by nzakas** [comment]
Sounds good.
